source/extensions/filters/http/grpc_field_extraction/message_converter/message_converter.cc

```diff
@@ -0,0 +1,159 @@
+
+  ENVOY_LOG_MISC(info, "len(parsing_buffer_)={}", parsing_buffer_.length());
+  if (parsed_output->owned_bytes != nullptr) {
+    ABSL_DCHECK(!parsed_output->needs_more_data);
+    ENVOY_LOG_MISC(info, "len(parsed owned_bytes)={}", parsed_output->owned_bytes->length());
+  }
+  return message_data;
+}
+
+absl::StatusOr<std::vector<std::unique_ptr<StreamMessage>>>
+MessageConverter::accumulateMessages(Envoy::Buffer::Instance& data, bool end_stream) {
+  std::vector<std::unique_ptr<StreamMessage>> messages;
+  while (true) {
+    absl::StatusOr<std::unique_ptr<StreamMessage>> message = accumulateMessage(data, end_stream);
+    if (!message.ok()) {
+      return message.status();
+    }
+    if (*message == nullptr) {
+      return messages;
+    }
+    messages.push_back(std::move(*message));
+    if (messages.back()->isFinalMessage()) {
+      return messages;
+    }
+  }
+}
+
+absl::StatusOr<Envoy::Buffer::InstancePtr>
+MessageConverter::convertBackToBuffer(std::unique_ptr<StreamMessage> message) {
+  ABSL_DCHECK(message != nullptr);
+  conversions_to_envoy_buffer_++;
+  if (conversions_to_envoy_buffer_ > conversions_to_message_data_) {
+    return absl::FailedPreconditionError(absl::StrCat(
+        "Data corruption! Number of conversions to StreamMessage = ", conversions_to_message_data_,
+        ", but this is the ", conversions_to_envoy_buffer_,
+        " convertBackToBuffer call. Perhaps this StreamMessage"
+        "belongs to a different MessageConverter?"));
+  }
+
+  // Edge case handling: If there is no proto body, nothing to convert.
+  Envoy::Buffer::InstancePtr output_message = std::make_unique<Envoy::Buffer::OwnedImpl>();
+  if (message->size() == -1) {
+    return output_message;
+  }
+
+  // Create gRPC frame header and add to output buffer.
+  const uint64_t out_message_size = message->size();
+  absl::StatusOr<std::string> delimiter = sizeToDelimiter(out_message_size);
+  if (!delimiter.ok()) {
+    return delimiter.status();
+  }
+  output_message->add(*delimiter);
+
+  std::unique_ptr<ZeroCopyInputStream> input_stream =
+      message->message()->CreateZeroCopyInputStream();
+
+  // Do NOT let StreamMessage and the current parsed data go out of scope and
+  // automatically destruct. Manage destruction lazily via lambda capture.
+  auto message_lifetime = std::shared_ptr<StreamMessage>(message.release());
+  auto fragment_releasor =
+      [message_lifetime](const void*, size_t,
+                         const Envoy::Buffer::BufferFragmentImpl* this_fragment) {
+        delete this_fragment;
+      };
+
+  // Data owned in StreamMessage is appended to Envoy's Buffer via external
+  // data reference (no copy).
+  const void* data = nullptr;
+  int size = 0;
+  bool is_empty = true;
+
+  while (input_stream->Next(&data, &size)) {
+    is_empty = false;
+    auto* last_fragment = new Envoy::Buffer::BufferFragmentImpl(data, size, fragment_releasor);
+    output_message->addBufferFragment(*last_fragment);
+  }
```

Comment:
IIRC we do similar thing in gRPC transcoding and/or other filters, consider refactor this to common protobuf util.

Reply:
Are you meaning reading chars from ZeroCopyStream?  Could you provide an example what you want it be? BTW, can I make it as a TODO?

Reply:
Ping @lizan 
BTW, the `envoy-presubmit (Prechecks Precheck dependencies)` seems not caused by my change. Could you help take a look?

Reply:
Similar things is here but copying. Let's open an issue as TODO to track then.